### ISSUE TITLE
Caching doesn't work: inconsistent results for pressure calculation

### ISSUE THREAD

**Issue by shuntstrato** [usnistgov/REFPROP-issues#543]
### Description

I have been having trouble getting results from REFPROPdll to match those of the GUI for certain calculations, including the pressure calculation demonstrated below. In this demonstration code, I haphazardly tinker with settings and rerun the same REFPROPdll command; certain sequences yield correct answers for this calculation, but these sequences don't seem to work for all calculations. I would appreciate advice on what settings to use to match the behavior of the REFPROP GUI, which has been yielding reliable results.

### Versions

**REFPROP Version:** 10
**Operating System and Version:** Windows 10
**Access Method:** Python 3.8.13

### Additional Information

Code example:

``` python
# =============================================================================
# initial setup
# =============================================================================

from ctREFPROP.ctREFPROP import REFPROPFunctionLibrary
import os

#props to stay constant for all calcs
mole_frac   = [0.4622481724404697, 0.5377518275595302]
e           = 4219.922120913738
rho         = 0.3726994434805592

#initialize refprop
RP = REFPROPFunctionLibrary(os.environ['RPPREFIX'])
RP.SETPATHdll(os.environ['RPPREFIX'])
units_code=RP.GETENUMdll(0, "MOLAR ENGLISH").iEnum
RP.FLAGSdll('Reset all', 2)

# =============================================================================
# first try will be wrong
# =============================================================================

P=RP.REFPROPdll('O2;N2O','ED','P',units_code, iMass=0,iFlag=1,
              a=e,b=rho, z=mole_frac).Output[0]
print(P)

# =============================================================================
# turn off caching, it will work now.
# =============================================================================

RP.FLAGSdll('Cache', 3) 

P=RP.REFPROPdll('O2;N2O','ED','P',units_code, iMass=0,iFlag=1,
              a=e,b=rho, z=mole_frac).Output[0]
print(P)

# =============================================================================
# reset everything. it will break again
# =============================================================================

RP.FLAGSdll('Reset all', 2)

P=RP.REFPROPdll('O2;N2O','ED','P',units_code, iMass=0,iFlag=1,
              a=e,b=rho, z=mole_frac).Output[0]
print(P)

# =============================================================================
# reset everything and turn cache off. still won't work.
# =============================================================================

RP.FLAGSdll('Reset all', 2)
RP.FLAGSdll('Cache', 3) 

P=RP.REFPROPdll('O2;N2O','ED','P',units_code, iMass=0,iFlag=1,
              a=e,b=rho, z=mole_frac).Output[0]
print(P)

# =============================================================================
# retry the case where we reset, fail a calc, turn off cache, and then redo
# the calc. it works!
# =============================================================================

RP.FLAGSdll('Reset all', 2)

P=RP.REFPROPdll('O2;N2O','ED','P',units_code, iMass=0,iFlag=1,
              a=e,b=rho, z=mole_frac).Output[0]
print(P)

RP.FLAGSdll('Cache', 3) 

P=RP.REFPROPdll('O2;N2O','ED','P',units_code, iMass=0,iFlag=1,
              a=e,b=rho, z=mole_frac).Output[0]
print(P)

```

Code results (the answer should be 1232 psi): 

836.2306729042862
1232.0002478847
836.2306729042862
836.2306729042961
836.2306729042862
1232.0002478847

**Comment by ianhbell** [comment]
Can you try the same exercise with SI units? I'm curious to see if that is part of the problem.

**Comment by ianhbell** [comment]
Nice report, by the way.

**Comment by shuntstrato** [comment]
Would you recommend DEFAULT, MOLAR SI, or MOLAR BASE SI? I prefer mass-based units, however I recall a previous post recommending molar-based.

**Comment by shuntstrato** [comment]
Here's a code in which I use MOLAR SI units to perform the same calculations as before:

```python
# =============================================================================
# initial setup
# =============================================================================

from ctREFPROP.ctREFPROP import REFPROPFunctionLibrary
import os


#props to stay constant for all calcs
mole_frac   = [0.4622481724404697, 0.5377518275595302]
e           = 4219.922120913738 * 1055.056 / 453.59237
rho         = 0.3726994434805592 * 0.0353146667214886 * 453.59237

#initialize refprop
RP = REFPROPFunctionLibrary(os.environ['RPPREFIX'])
RP.SETPATHdll(os.environ['RPPREFIX'])
units_code=RP.GETENUMdll(0, "MOLAR SI").iEnum
RP.FLAGSdll('Reset all', 2)

# =============================================================================
# first try
# =============================================================================

P=RP.REFPROPdll('O2;N2O','ED','P',units_code, iMass=0,iFlag=1,
              a=e,b=rho, z=mole_frac).Output[0]
print(P*145.03773773020922)

# =============================================================================
# turn off caching
# =============================================================================

RP.FLAGSdll('Cache', 3) 

P=RP.REFPROPdll('O2;N2O','ED','P',units_code, iMass=0,iFlag=1,
              a=e,b=rho, z=mole_frac).Output[0]
print(P*145.03773773020922)


# =============================================================================
# reset everything
# =============================================================================

RP.FLAGSdll('Reset all', 2)

P=RP.REFPROPdll('O2;N2O','ED','P',units_code, iMass=0,iFlag=1,
              a=e,b=rho, z=mole_frac).Output[0]
print(P*145.03773773020922)


# =============================================================================
# reset everything and turn cache off
# =============================================================================

RP.FLAGSdll('Reset all', 2)
RP.FLAGSdll('Cache', 3) 

P=RP.REFPROPdll('O2;N2O','ED','P',units_code, iMass=0,iFlag=1,
              a=e,b=rho, z=mole_frac).Output[0]
print(P*145.03773773020922)


# =============================================================================
# retry the case where we reset, fail a calc, turn off cache, and then redo
# the calc.
# =============================================================================

RP.FLAGSdll('Reset all', 2)

P=RP.REFPROPdll('O2;N2O','ED','P',units_code, iMass=0,iFlag=1,
              a=e,b=rho, z=mole_frac).Output[0]
print(P*145.03773773020922)


RP.FLAGSdll('Cache', 3) 

P=RP.REFPROPdll('O2;N2O','ED','P',units_code, iMass=0,iFlag=1,
              a=e,b=rho, z=mole_frac).Output[0]
print(P*145.03773773020922)
```

The results are sort of flipped in this case. Converted back to psi, they are: 

1234.2847992979082
4302.751296855292
1234.2847992979082
1234.2847992981262
1234.2847992979082
4302.751296855292

So in this case, the default settings yielded the correct-ish answer. However: I suspect my conversions introduced some error in the inputs. An interesting thing I noticed about the previous (English-unit) code but did not report is that when I rounded _e_ to 4220, it output:

1232.0629904389348
4298.461105619768
1232.0629904389348
1232.062990439175
1232.0629904389348
4298.461105619768

To summarize: I know that with English units, miniscule input changes flip which settings yield correct and incorrect results. I suspect that we may be seeing the same behavior with SI, and conversion error led to the flipping of the results.

**Comment by ianhbell** [comment]
@EricLemmon Any idea what is going on? Maybe a caching bug?

**Comment by Ranzengar** [comment]
Hi @ianhbell have you solved this issue? We are having the same problem:
For the same input we have different results despite forcing a cache reset using either 'Reset All' or 'Cache' props or the two together.
It seems a caching problem to me. There is some other way to tell to the dll to not cache anything? I'm not worried about performance issues at the moment, i just need it to be reliable

**Comment by ianhbell** [comment]
In theory, setting Cache to 3 should disable caching: https://refprop-docs.readthedocs.io/en/latest/DLL/high_level.html#f/_/FLAGSdll

**Comment by Ranzengar** [comment]
Yes but it still doesn't work sometimes, as the example in this thread

**Comment by ianhbell** [comment]
We need @EricLemmon to chime in here. Do you want to try the beta? There is a small chance it will help. If so, email me at ian.bell@nist.gov

**Comment by EricLemmon** [comment]
Can you try the beta first before I look at this?  It fixes a lot of these strange bugs.
